#630 transmute_vec_as_bytes potential fix

NathanaelG1 opened the pull request:
Hey all new here but believe this may fix the issue raised in #630. Added some comments on the issue thread including a playground which shows the result of the old function versus the new function. I have taken an approach to only account for the specific types which actually are used in the codebase (f32 and usize) as there were a lot of issues dealing with a totally generic type. I understand this adds a limitation that was previously not there, if this is not a good approach let me know and I can do some reformatting. 

I have added two unit tests for transmute_vec_as_bytes which makes sure the results do not include any random values from the padding. 

I removed the need for an unsafe block because the types are known to the function, and did not add any dependencies to the project.

All unit tests are passing, and running my modified tutorials from the RPG and FPS on the documentation seem to run the same as on the master branch of the main fork. This is my first PR here if there's anything I can do to improve it let me know!

mrDIMAS commented:
Hi, this looks like a partial solution. May I ask you, could you please try using `bytemuck` crate for this case? Afaik all you need to do is to add `Pod` trait bound to `T` and this should automagically fix the issue.

NathanaelG1 commented:
@mrDIMAS 

I have added bytemuck and added the pod trait to T, on compiling there is an error that Pod should be implemented for BlendShapesContainer::from_lists::VertexData. When I implement it there using the derive macro there are 3 more errors for every Vec3 declared in this impl block. 

Using bytemuck may require a lot more changes to the code I haven't gotten through them all yet, minimally bytemuck will need to be added to fyrox-core and fyrox-impl and a pod wrapper with implementations of Pod and Zeroable to be added to Vertex Data. If this approach sounds good I'll continue.

NathanaelG1 commented:
[@mrDIMAS](https://github.com/mrDIMAS)

I have added `bytemuck` and applied the `Pod` trait to `T`. During compilation, an error indicated that `Pod` should be implemented for `BlendShapesContainer::from_lists::VertexData`. When I attempted to derive `Pod` for `VertexData`, additional errors surfaced for each `Vec3` declared in the implementation block.

Using `bytemuck` might require significant changes to the codebase. At a minimum, `bytemuck` would need to be added to both `fyrox-core` and `fyrox-impl`, and a `Pod` wrapper with implementations of `Pod` and `Zeroable` would need to be created for `VertexData`.

If this approach sounds acceptable, I'll proceed with the necessary changes. 

errors: 
   --> fyrox-impl/src/scene/mesh/surface.rs:170:21
    |
170 |         let bytes = crate::core::transmute_vec_as_bytes(vertex_data);
    |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot infer type of the type parameter `Pod` declared on the function `transmute_vec_as_bytes`
    |

   --> fyrox-impl/src/scene/terrain/mod.rs:116:9
    |
116 |         crate::core::transmute_vec_as_bytes(height_map),
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot infer type of the type parameter `Pod` declared on the function `transmute_vec_as_bytes`
    |

mrDIMAS commented:
Yes, you can proceed. As for `nalgebra` items (Vectors, etc) - you can add `bytemuck` feature to `nalgebra` crate, and it will implement the required traits.

NathanaelG1 commented:
Made the suggested changes and left the unit tests.

mrDIMAS commented on fyrox-impl/src/scene/mesh/surface.rs:
```diff
@@ -94,18 +95,45 @@ pub struct InputBlendShapeData {
     pub tangents: FxHashMap<u32, Vector3<f16>>,
 }
 
+#[repr(C)]
+#[derive(Default, Clone, Copy)]
+struct PodVector3 {
+    x: f16,
+    y: f16,
+    z: f16,
+}
+
+unsafe impl Pod for PodVector3 {}
+unsafe impl Zeroable for PodVector3 {}
+
+impl From<Vector3<f16>> for PodVector3 {
+    fn from(v: Vector3<f16>) -> Self {
+        PodVector3 {
+            x: v.x,
+            y: v.y,
+            z: v.z,
+        }
+    }
+}
+
+impl From<PodVector3> for Vector3<f16> {
+    fn from(pv: PodVector3) -> Self {
+        Vector3::new(pv.x, pv.y, pv.z)
+    }
+}
+
```
`half` crate has `bytemuck` feature as well, so this `PodVector3` can be replaced by simply enabling the `bytemuck` feature.
